contrib/vcloud/podman_containerized_tool.py

```diff
@@ -0,0 +1,163 @@
+
+
+def _init_container(
+    image,
+    tool_directory,
+):
+    """
+    Move this process into a container.
+    """
+    volumes = []
+
+    tool_directory = os.path.abspath(tool_directory)
+
+    # Mount the python loaded paths into the container
+    # The modules are mounted at the exact same path in the container
+    # because we do not yet know a solution to tell python to use
+    # different paths for the modules in the container.
+    python_paths = [path for path in sys.path if os.path.isdir(path)]
+    for path in python_paths:
+        abs_path = os.path.abspath(path)
+        volumes.extend(["--volume", f"{abs_path}:{abs_path}:ro"])
+
+    # Mount the tool directory into the container at a known location
+    volumes.extend(
+        [
+            "--volume",
+            f"{tool_directory}:{TOOL_DIRECTORY_MOUNT_POINT}:O",
+            # :O creates an overlay mount. The tool can write files in the container
+            # but they are not visible outside the container.
+            "--workdir",
+            "/mnt",
+        ]
+    )
+
+    # Create a container that does nothing but keeps running
+    command = (
+        ["podman", "run", "--entrypoint", "tail", "--rm", "-d"]
+        + volumes
+        + [image, "-F", "/dev/null"]
+    )
+
+    logging.debug(
+        "Command to start container: %s",
+        shlex.join(map(str, command)),
+    )
+    res = subprocess.run(
+        command,
+        stdout=subprocess.PIPE,
+    )
+
+    res.check_returncode()
+    container_id = res.stdout.decode().strip()
+
+    container_pid = (
+        subprocess.run(
+            ["podman", "inspect", "--format", "{{.State.Pid}}", container_id],
+            stdout=subprocess.PIPE,
+        )
+        .stdout.decode()
+        .strip()
+    )
+
+    def join_ns(namespace):
+        namespace = f"/proc/{container_pid}/ns/{namespace}"
+        logging.debug("Joining namespace %s .", namespace)
+        with open(namespace, "rb") as f:
+            libc.setns(f.fileno(), 0)
+
+    try:
+        logging.debug("Joining namespaces of container %s.", container_id)
+
+        necessary_namespaces = frozenset(("user", "mnt"))
+
+        # The user namespace must be joined first
+        # because the other namespaces depend on it
+        join_ns("user")
+
+        for namespace in os.listdir(f"/proc/{container_pid}/ns"):
+            if namespace in necessary_namespaces:
+                continue
+            try:
+                # We try to mount all listed namespaces, but some might not be available
+                join_ns(namespace)
+
+            except OSError as e:
+                logging.debug(
+                    "Failed to join namespace %s: %s", namespace, os.strerror(e.errno)
+                )
+
+        # The mount namespace must be joined so we want
+        # to fail if we cannot join the mount namespace.
+        # mnt must be joined last because after joining it,
+        # we can no longer access /proc/<container_pid>/ns
```

Comment:
This could actually be solved by opening `/proc/<container_pid>/ns` as a directory fd and making all subsequent accesses relative to that fd. Not sure if it makes things easier, but it is possible.